### PR TITLE
Test multiple CMake versions in Travis CI using docker

### PR DESCRIPTION
**Pull request by diegoferigo**
Implemented features:
* Testing of CMake >= 3.1 on Linux environment
* Tests on `rc` versions are allowed to fail
* CMake archives caching
* Linux generators:
  * Unix Makefiles
  * Ninja
* Osx generators:
  * Unix Makefiles
  * Xcode

TODO and comments:
* CMake 3.0 has only `i386` tarballs and it would require the installation of `lib32` packages. Since this version will be deprecated sometime soon, I didn't add this exception
* In osx, only CMake 3.6.2 (the one shipped inside Travis' virtualization environment) is tested. The support of multiple versions is already present in the build matrix, but I miss macOS knowledge to properly handle packages.

### CODE REVIEW

**Comment by diegoferigo** [comment]
@drdanz can you please give me some feedback about [these](https://travis-ci.org/robotology/ycm/jobs/215426102#L686) tests that are failing? Even testing against against `ycm:master` they keep failing.

P.s. In the ctest commit I moved that section inside the `script` step, and the build is hence failing. What's the point of having them in the `after_script` step as in the current `.travis.yml`?

File: .travis.yml [hunk 1]
```diff
@@ -1,41 +1,201 @@
-sudo: required
+sudo: false
 dist: trusty
+osx_image: xcode7.3
```
Comment (drdanz) [inline .travis.yml]: Please add this to the xcode images only (see [here](https://github.com/robotology/yarp/blob/devel/.travis.yml#L111)), I find it very annoying that the output matrix shows this for all the linux lines

**Comment by drdanz** [comment]
> CMake 3.0 has only i386 tarballs and it would require the installation of lib32 packages. Since this version will be deprecated sometime soon, I didn't add this exception

Actually this is the most critical one since it lacks of a few features, if it is possible to do that, I'd be quite happy 😉

File: .travis.yml (hunk 1 above)
Comment (diegoferigo) [inline .travis.yml]: Done

**Comment by diegoferigo** [comment]
> Actually this is the most critical one since it lacks of a few features, if it is possible to do that, I'd be quite happy  :wink:

It was easier than expected, CMake does not require several lib32. Anyway, CMake 3.0 doesn't succeed the configuration step, there are issues with `CONFIGURE_PACKAGE_CONFIG_FILE`.

> The output is not very useful, can you set the environment variable CTEST_OUTPUT_ON_FAILURE=1 (and ensure that it is forwarded to docker when you run the tests)?

Done, still not very helpful.

**Comment by diegoferigo** [comment]
In either `devel` and `master` branches, excluding CMake 3.0, these tests are failing:
```
5 - YCMBootstrap-not-use-system (Failed)
6 - YCMBootstrap-disable-find (Failed)
```

**Comment by drdanz** [comment]
> CMake 3.0 doesn't succeed the configuration step, there are issues with CONFIGURE_PACKAGE_CONFIG_FILE.

Great! 😭 
You can add it to the allowed failures for now...

**Comment by drdanz** [comment]
> In either `devel` and `master` branches, excluding CMake 3.0, these tests are failing:
> ```
> 5 - YCMBootstrap-not-use-system (Failed)
> 6 - YCMBootstrap-disable-find (Failed)
> ```

It looks like something related to docker, the master and devel branches don't seem to have issues at the moment... Anyway I just restarted them ([[master](https://travis-ci.org/robotology/ycm/builds/195886251)], [[devel](https://travis-ci.org/robotology/ycm/builds/210748200)]), we'll see when they are finished...

**Comment by drdanz** [comment]
Apparently they are working... https://travis-ci.org/robotology/ycm/jobs/195886252#L828

**Comment by diegoferigo** [comment]
> You can add it to the allowed failures for now...

Ok!

> It looks like something related to docker, the master and devel branches don't seem to have issues at the moment... Anyway I just restarted them ([master], [devel]), we'll see when they are finished...

Yes, `master` has some issues I still not understanding. But the [upstream devel branch](https://travis-ci.org/robotology/ycm/jobs/210748201#L766) has issues too, the build passes only because ctest is in the `after_script` section.

**Comment by drdanz** [comment]
Hmm interesting, I just noticed that the `devel` branch is failing, but it does not report the error...
Anyway, there is a `tools/update_sha1sums.sh` that regenerate the sha1sums, you can try executing that and committing the modified files...

**Comment by drdanz** [comment]
Actually I just did it :smile:

**Comment by drdanz** [comment]
@diegoferigo Do you remember what was blocking this patch? I forgot...

**Comment by diegoferigo** [comment]
@drdanz Yes, I remember few tests were failing when run inside docker. I never had time to have a deeper look.

**Comment by drdanz** [comment]
In this case I will just close this, we can re-open it when we have time to work on it...

**Pull request closed without merging**